### PR TITLE
GH-35282: [C++] auto enable brotli when enable fuzzing

### LINKED ISSUE #35282: [C++] auto enable brotli when enable ARROW_FUZZING
### Describe the enhancement requested

In `parquet/arrow/generate_fuzz_corpus.cc`, it depends on brotli. However, arrow didn't auto enable it:

```C++
std::shared_ptr<WriterProperties> GetWriterProperties() {
  WriterProperties::Builder builder{};
  builder.disable_dictionary("no_dict");
  builder.compression("compressed", Compression::BROTLI);
  return builder.build();
}
````

### Component(s)

C++, Developer Tools

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

Auto enable brotli when enable fuzzing

### What changes are included in this PR?

Detect Fuzzing enabled, if it's, enable BROTLI

### Are there any user-facing changes?

* Closes: #35282

### CODE REVIEW

**Review by felipecrv — APPROVED** [review]
LGTM.

**Comment by mapleFU** [comment]
@wjones127 @kou Would you mind take a look?

**Comment by kou** [comment]
Could you do this in `cpp/cmake_modules/DefineOptions.cmake`?

```diff
diff --git a/cpp/cmake_modules/DefineOptions.cmake b/cpp/cmake_modules/DefineOptions.cmake
index bbbcbe7eb..a4f7a7871 100644
--- a/cpp/cmake_modules/DefineOptions.cmake
+++ b/cpp/cmake_modules/DefineOptions.cmake
@@ -264,7 +264,8 @@ takes precedence over ccache if a storage backend is configured" ON)
                 "Build Arrow Fuzzing executables"
                 OFF
                 DEPENDS
-                ARROW_TESTING)
+                ARROW_TESTING
+                ARROW_WITH_BROTOLI)
 
   define_option(ARROW_LARGE_MEMORY_TESTS "Enable unit tests which use large memory" OFF)
 
```

**Comment by mapleFU** [comment]
Ooops... CI Boomed now?

File: cpp/cmake_modules/DefineOptions.cmake [hunk 1]
```diff
@@ -264,7 +264,8 @@ takes precedence over ccache if a storage backend is configured" ON)
                 "Build Arrow Fuzzing executables"
                 OFF
                 DEPENDS
-                ARROW_TESTING)
+                ARROW_TESTING
+                ARROW_WITH_BROTLI)
```
Comment (felipecrv) [inline cpp/cmake_modules/DefineOptions.cmake:268]: This is nice, but it means build scripts will error out when options don't satisfy these dependencies, right?

Comment (kou) [reply cpp/cmake_modules/DefineOptions.cmake:268]: It's not right.
It means that "-DARROW_WITH_BROTLI=ON is added implicitly when -DARROW_FUZZING=ON is specified".
We don't need to update the CI scripts .

**Comment by kou** [comment]
The Python failures will be fixed by #35324.
I think that the R failure is unrelated but I re-ran the job.

**Comment by kou** [comment]
The R failure is happen on main too: https://github.com/apache/arrow/actions/runs/4803502649/jobs/8548049946

**Review by kou — APPROVED** [review]
+1

**Pull request merged**